ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/DefaultBlockchain.java

```diff
@@ -127,40 +127,69 @@ private DefaultBlockchain(
+    if (headersCacheSize == 0 && blocksCacheSize == 0) {
+      setAllCachesEmpty();
+      return;
+    }
+
+    int headersSize =
+        (headersCacheSize != 0) ? Math.max(headersCacheSize, blocksCacheSize) : blocksCacheSize;
+    blockHeadersCache =
+        Optional.of(CacheBuilder.newBuilder().recordStats().maximumSize(headersSize).build());
+    ;
+
+    if (blocksCacheSize != 0) {
       blockBodiesCache =
-          Optional.of(
-              CacheBuilder.newBuilder().recordStats().maximumSize(numberOfBlocksToCache).build());
+          Optional.of(CacheBuilder.newBuilder().recordStats().maximumSize(blocksCacheSize).build());
+      ;
       transactionReceiptsCache =
-          Optional.of(
-              CacheBuilder.newBuilder().recordStats().maximumSize(numberOfBlocksToCache).build());
+          Optional.of(CacheBuilder.newBuilder().recordStats().maximumSize(blocksCacheSize).build());
+      ;
       totalDifficultyCache =
-          Optional.of(
-              CacheBuilder.newBuilder().recordStats().maximumSize(numberOfBlocksToCache).build());
+          Optional.of(CacheBuilder.newBuilder().recordStats().maximumSize(blocksCacheSize).build());
+      ;
       blockAccessListCache =
-          Optional.of(
-              CacheBuilder.newBuilder().recordStats().maximumSize(numberOfBlocksToCache).build());
-      metricsSystem.createGuavaCacheCollector(BLOCKCHAIN, "blockHeaders", blockHeadersCache.get());
-      metricsSystem.createGuavaCacheCollector(BLOCKCHAIN, "blockBodies", blockBodiesCache.get());
-      metricsSystem.createGuavaCacheCollector(
-          BLOCKCHAIN, "transactionReceipts", transactionReceiptsCache.get());
-      metricsSystem.createGuavaCacheCollector(
-          BLOCKCHAIN, "totalDifficulty", totalDifficultyCache.get());
+          Optional.of(CacheBuilder.newBuilder().recordStats().maximumSize(blocksCacheSize).build());
+      ;
+      registerCacheMetrics(metricsSystem);
     } else {
-      blockHeadersCache = Optional.empty();
-      blockBodiesCache = Optional.empty();
-      transactionReceiptsCache = Optional.empty();
-      totalDifficultyCache = Optional.empty();
-      blockAccessListCache = Optional.empty();
+      // Only headers cache is created, rest are empty
+      setBlockCachesEmpty();
+      registerHeadersCacheMetrics(metricsSystem);
     }
+  }
 
-    createCounters(metricsSystem);
-    createGauges(metricsSystem);
+  private void setAllCachesEmpty() {
+    blockHeadersCache = Optional.empty();
+    setBlockCachesEmpty();
+  }
+
+  private void setBlockCachesEmpty() {
+    blockBodiesCache = Optional.empty();
+    transactionReceiptsCache = Optional.empty();
+    totalDifficultyCache = Optional.empty();
+    blockAccessListCache = Optional.empty();
+  }
+
+  private void registerCacheMetrics(final MetricsSystem metricsSystem) {
+    metricsSystem.createGuavaCacheCollector(BLOCKCHAIN, "blockHeaders", blockHeadersCache.get());
```

Comment:
maybe call the other method here, for uniformity this the way empty caches are set?

```suggestion
    registerHeadersCacheMetrics(metricsSystem);
```